File: Lab1/prob1.cpp
#include <stdio.h>
#include <iostream>
using namespace std;

#define N 5

bool prefer(int n, int w, int m, int curr, int w_pref[N][5]){
	for (int i=0;i<n;i++){
		if (w_pref[w][i]==m)
			return true;
        else
         return false;

	}

}

int main(int argc, char const *argv[])
{
	int n;
	cout<<"Enter the value of n: ";
//	cin>>n;
n=N;
	int men[n],women[n],m_pref[n][n],w_pref[n][n];
	for (int i=0;i<n;i++){
		men[i]=i;
		women[i]=i;
	}

	for (int i=0;i<n;i++){
		cout<<"Enter the preference of man "<<i<<" ";
		for (int j=0;j<n;j++){
			cin>>m_pref[i][j];
		}
	}

	for (int i=0;i<n;i++){
		cout<<"Enter the preference of woman "<<i<<" ";
		for (int j=0;j<n;j++){
			cin>>w_pref[i][j];
		}
	}

	bool mFree[n];
	int husband[n];

	for (int i=0;i<n;i++){
		husband[i]=-1;
		mFree[i]=false;
	}

	int free=n;
	int curr,j,i;


	while(free>0){

		for (j=0;j<n;j++){
			if (mFree[j]==false)
				break;
		}

		for (i=0;i<n;i++){
			int w=m_pref[j][i];

			if (husband[w]!=-1){

				husband[w]=j;
				mFree[j]=false;
				free--;
			}

			else{
				curr=husband[w];

				if(prefer(n,w,j,curr,w_pref[][5])==true){

					husband[w]=j;
					mFree[j]=false;
					mFree[curr]=true;


				}
			}
		}
	}

	cout<<"Woman 		Man"<<'\n';
	for(int i=0;i<n;i++)
		cout<<i+1<<" 		"<<husband[i]<<endl;

	return 0;
}

File: Lab1/prob1.py

def wPrefer(w_pref,j,m1,w):
	for i in range(len(w_pref)):
		if w_pref[w][i] == j:
			print(w,"changes from man ",m1," to ",j )
			return True
		if w_pref[w][i]==m1:
			return False	


#n=int(input("Enter n: "))
n=3
m_pref=[[0 for i in range(n)] for k in range(n)]
w_pref=[[0 for i in range(n)] for k in range(n)]


# lines=[line.rstrip('\n') for line in open('input.txt')]
# k=0
# # print(lines[1].strip(" "))
# print(lines[2][4])
# for i in range(2,2+n):
# 	for j in range(2,n,2):
# 		m_pref[i-2][k]=int(lines[i][j])
# 		k+=1
# k=0
# for i in range(2+n,2+2*n):
# 	for j in range(2,n,2):
# 		w_pref[i-n-2][k]=int(lines[i][j])
# 		k+=1

# print(m_pref[0][1])

Men=['m1','m2','m3']
Women=['w1','w2','w3']
m_pref=[[1, 2, 0],
		[1, 0, 2],
		[0, 2, 1]]
w_pref=[[1, 2, 0],[1,0, 2],[2, 1, 0]]

# for i in range(n):
# 	print("Enter preference of man "+str(i)+": ")
# 	for j in range(n):
# 		m_pref[i][j]=int(input(""))

# for i in range(n):
# 	print("Enter preference of woman "+str(i)+": ")
# 	for j in range(n):
# 		w_pref[i][j]=int(input(""))

husband=[None for i in range(n)]
wife=[None for i in range(n)]
mFree=[False for i in range(n)]
free=n


while free>0:
	for j in range(n):
		if mFree[j]==False:
			break

	for i in range(n):
		if mFree[j]==False:
			w=m_pref[j][i]

			if husband[w]==None:
				husband[w]=j
				wife[j]=w

				mFree[j]=True
				free-=1
			else:
				m1=husband[w]

				if wPrefer(w_pref,j,m1,w)==True:
					husband[w]=j
					wife[j]=w
					mFree[j]=True
					mFree[m1]=False

print("Man\tWoman")
for i in range(0,n):
	print(str(Men[i])+"\t"+Women[wife[i]])


